### PR TITLE
Br v4.0.0

### PR DESCRIPTION
**Pull request by annakrystalli**
This PR:

* BREAKING CHANGE: introduces `is_required` boolean property at the `output_type` level to configure whether the output type is required for submissions to be considered valid (#99). 
* BREAKING CHANGE: Disallows `optional` property in `output_type_id` objects. As such, when a given output type is submitted, values for all output type IDs much be submitted (#100,#101, #102). 
* Introduces optional `derived_task_ids` property to enable hub administrators to define derived task IDs (i.e. task IDs whose values depend on the values of other task IDs) at a hub level. This allows primarily validation functionality to ignore such task IDs when appropriate which can significantly improve validation efficency (#96 ). For more information see [`hubValidations` documentation on ignoring derived task IDs](https://hubverse-org.github.io/hubValidations/articles/validate-pr.html#ignoring-derived-task-ids-to-improve-performance).
* Add descriptions to repository object (#98)
* Also clarified branch naming instruction to align with expectation in `hubDocs`

### CODE REVIEW

**Comment by annakrystalli** [comment]
\diff

**Comment by annakrystalli** [comment]
### `admin-schema.json`

```diff
39,43c45,51
<             "examples": [{
<                 "host": "github",
<                 "owner": "hubverse-org",
<                 "name": "example-simple-forecast-hub"
<             }],
---
>             "examples": [
>                 {
>                     "host": "github",
>                     "owner": "hubverse-org",
>                     "name": "example-simple-forecast-hub"
>                 }
>             ],
49c57,59
<                     "enum": ["github"]
---
>                     "enum": [
>                         "github"
>                     ]
51a62
>                     "description": "The hub repository owner (user or organisation).",
54a66
>                     "description": "The name of the hub repository.",
```

**Comment by annakrystalli** [comment]
Note I did manage to re-indent with `jq` but it expanded all arrays to multiple rows too with no clear way how to avoid that. I've tried to manually focus the diff on what's important in v4.0.0 instead.

If someone has better ideas let me know!

**Comment by zkamvar** [comment]
> Note I did manage to re-indent with jq but it expanded all arrays to multiple rows too with no clear way how to avoid that.

I think we should just bend to the formatter. One thing we could do to avoid this is to run 3.0.1 through the formatter in this PR as well so that the diff appears the same. The formatter will not change any structural information of the JSON, just its format. 

From there, we can add guidance for future additions to run it through the formatter.

**Review by LucieContamin — COMMENTED** [review]
Thank you for the update, I just added a question about the `derived_tasks_ids`.

**Comment by annakrystalli** [comment]
> I think we should just bend to the formatter. One thing we could do to avoid this is to run 3.0.1 through the formatter in this PR as well so that the diff appears the same. The formatter will not change any structural information of the JSON, just its format.

Good option! See #106 

> From there, we can add guidance for future additions to run it through the formatter.

I opened the following issue to discuss approach: #107

File: v4.0.0/tasks-schema.json [hunk 1]
```diff
@@ -0,0 +1,1496 @@
...
+        "description": "Names of derived task IDs, i.e. task IDs whose values are derived from (and therefore dependent on) the values of other variables.",
+        "type": [
+            "array",
+            "null"
+        ],
+        "uniqueItems": true,
+        "items": {
+            "type": "string"
+        }
+    },
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply v4.0.0/tasks-schema.json:1490]: I feel like it's a hub property overall in that I can't see many situations where a task ID is derived in one round and not derived in another. The chances of it being stable are far greater than the property changing by round so it would be annoying to have to re-define it again in every round. 

I'll make sure however that if a new derived task ID is added to a new round and the task id name added to `derived_task_ids`, it will not affect validation of older rounds, i.e. if the task ID is not present in the data, `derived_task_ids` will just be ignored.

The is the potential to allow for a  `derived_task_ids` property at the  round level that overrides the overall property but I think we can wait and see if anyone requests that as it feels like an extreme edge case?

[the remark this answers is not in the document]
Comment (LucieContamin) [reply v4.0.0/tasks-schema.json:1490]: Thank you for the additional information. 

You make a really good point, however I am worrying a bit about the scenario hubs. They tend to have lot of rounds and have different tasks id for some round as for the need of a specific round or scenario you might need to create new column that are not replicated in the next round. 
I am not sure how frequent this new columns will be tagged as derived tasks ids, but even saying that it looks to me a little bit weird to have the tasks id as hub level as they are not define at hub level but at round level.

[the remark this answers is not in the document]
Comment (annakrystalli) [reply v4.0.0/tasks-schema.json:1490]: I definitely don't want everyone to have to respecify this property at each round as for the majority of hubs this is very stable so we will for sure keep the overall specification. As mentioned there is the option for overriding at the round but I do not want to implement unless its actually necessary

[the remark this answers is not in the document]
Comment (LucieContamin) [reply v4.0.0/tasks-schema.json:1490]: > The problem though does not arise from a derived task id not being in all rounds 

I misunderstood and I was afraid that would be an issue but if not, that solve some of my problems! Apologize for that! 
I agree that changing the behavior of a column is bad practice and should not be supported.

> I definitely don't want everyone to have to respecify this property at each round as for the majority of hubs this is very stable so we will for sure keep the overall high level specification.

I also agree, that's to much to ask for something that might not be used often. 
And the optional overriding seems like a good idea,  and I think a hub can do their own wrapper/function to deal with it if necessary.

So to summarize, thank you for the additional information, I change my mind and I am ok with it being at hub level!

**Comment by annakrystalli** [comment]
\diff

**Comment by annakrystalli** [comment]
\diff

**Comment by annakrystalli** [comment]
Hey @LucieContamin ! I added a round level property now too so you can now use that to override the global property.

Glad you asked for this feature as going back to introduce it I found that the global `derived_task_ids` property was nested at the wrong level!!  🙈

**Comment by zkamvar** [comment]
## point estimates

Something that was brought up in response to https://github.com/reichlab/variant-nowcast-hub/pull/117#issuecomment-2423170370 is that the `"NA"` is a bit confusing because it sure looks like a character, but when we expand the grid the `output_type_id` columns become `NA` (which is an intentional move by Ooms described in [section 2.1.1 of the JSONlite package paper](https://cran.r-project.org/web/packages/jsonlite/vignettes/json-mapping.pdf))

Now that we are using `is_required` for point estimate types, **we might be able to take this opportunity to set the `required` property to a single element `null` array.** This will have exactly the same result as the `"NA"` array, but with the following advantages:

1. inter-language compatibility can be achieved since `null` is a concept that even JSON can understand
2. we can clearly communicate that this should be a missing value in the data as opposed to a character string. 

This is what I think it would look like in the schema: 

```
"required": {
    "description": "Not relevant for point estimate output types. Must be a single array of null"
    "type": "array"
    "items": {
        "const": null,
        "maxItems": 1
    }
}
```

### Demo

Here's a demo that shows that `["NA"]` and `[null]` are equivalent by modifying a tasks.json file and reading them in with jsonlite

``` r
hub_con <- hubData::connect_hub(
  system.file("testhubs/simple", package = "hubUtils")
)
hp <- attributes(hub_con)$hub_path
# Get the tasks file which has a `mean` output type
tasks <- fs::path(hp, "hub-config", "tasks.json")

# copy the file to a new temp file
tmp <- withr::local_tempfile()
fs::file_copy(tasks, tmp, overwrite = TRUE)
# The two files are identical
unname(tools::md5sum(tmp) == tools::md5sum(tasks))
#> [1] TRUE

cfg <- readLines(tmp)
writeLines(cfg[231:240]) # optional is NA
#>                 "output_type": {
#>                     "mean": {
#>                         "output_type_id": {
#>                             "required": null,
#>                             "optional": ["NA"]
#>                         },
#>                         "value": {
#>                             "type": "integer",
#>                             "minimum": 0
#>                         }

# Change '["NA"]' to '[null]'
nullcfg <- sub('["NA"]', '[null]', cfg, fixed = TRUE)
writeLines(nullcfg[231:240]) # optional is now null
#>                 "output_type": {
#>                     "mean": {
#>                         "output_type_id": {
#>                             "required": null,
#>                             "optional": [null]
#>                         },
#>                         "value": {
#>                             "type": "integer",
#>                             "minimum": 0
#>                         }
writeLines(nullcfg, con = tmp) # changing to null

# The two files are no longer identical
unname(tools::md5sum(tmp) == tools::md5sum(tasks))
#> [1] FALSE
waldo::compare(cfg, nullcfg)
#> old[81:87] vs new[81:87]
#>   "                    \"mean\": {"
#>   "                        \"output_type_id\": {"
#>   "                            \"required\": null,"
#> - "                            \"optional\": [\"NA\"]"
#> + "                            \"optional\": [null]"
#>   "                        },"
#>   "                        \"value\": {"
#>   "                            \"type\": \"integer\","
#> 
#> old[232:238] vs new[232:238]
#>   "                    \"mean\": {"
#>   "                        \"output_type_id\": {"
#>   "                            \"required\": null,"
#> - "                            \"optional\": [\"NA\"]"
#> + "                            \"optional\": [null]"
#>   "                        },"
#>   "                        \"value\": {"
#>   "                            \"type\": \"integer\","


# The resulting R object after reading in JSON are identical
identical(
  jsonlite::fromJSON(tasks, simplifyDataFrame = FALSE), 
  jsonlite::fromJSON(tmp, simplifyDataFrame = FALSE)
)
#> [1] TRUE
```

<sup>Created on 2024-10-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>
